openedx/openedx-platform
# [BB-933] Allow overriding xblock icons with plugins

This is a follow-up to #19772, which allowed setting custom icons for each unit in Studio. By default the icon would be derived from unit's type. The suggestion was to do move it to a plugin. This PR introduces some changes required to move it to external Django application

<s>Most of this diff is about extracting modals from monolithic `cms/static/js/views/modals/course_outline_modals.js` to make them reusable. We also extracted `CourseOutlineView` class into a field in `cms/static/js/views/pages/course_outline.js` to make this view easily extendable.</s>

For overriding implementation of getting icon, we implemented `pluggable_override` decorator.

**JIRA tickets**: [OSPR-3801](https://openedx.atlassian.net/browse/OSPR-3801)

**Merge deadline**: None

**Sandbox**: https://pr21433.sandbox.opencraft.hosting/ (provisioning)

**Testing instructions**:

1. Go to [this page](http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/courseware/interactive_demonstrations/19a30717eff543078a5d94ae9d6c18a5/?child=first) and see that the first unit icon is [fa-book](https://fontawesome.com/v4.7.0/icon/book).
1. Install the `master` version of [edx-django-utils](https://github.com/edx/edx-django-utils) in LMS (`make lms-shell`):
    `pip install -e git+https://github.com/edx/edx-django-utils.git#egg=edx-django-utils`
1. Install [custom-unit-icons](https://github.com/open-craft/custom-unit-icons) for LMS (`make lms-shell`):
    `pip install -e git+https://github.com/open-craft/custom-unit-icons.git@agrendalath/bb-933-address_override_changes#egg=custom-unit-icons`
1. Set the following in `lms/envs/private.py` and run `make lms-restart`:
   ```
   from .common import XBLOCK_MIXINS
   OVERRIDE_GET_UNIT_ICON = 'custom_unit_icons.icons.get_icon'
   XBLOCK_MIXINS += ('custom_unit_icons.icons.IconOverrideMixin',)
   ```
1. Go to [this page](http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/courseware/interactive_demonstrations/19a30717eff543078a5d94ae9d6c18a5/?child=first) and see that the first unit icon is still [fa-book](https://fontawesome.com/v4.7.0/icon/book).
1. Run the following snippet in LMS Django shell:
   ```python
   from xmodule.modulestore.django import modulestore
   from opaque_keys.edx.keys import UsageKey

   usage_key = UsageKey.from_string('block-v1:edX+DemoX+Demo_Course+type@vertical+block@867dddb6f55d410caaa9c1eb9c6743ec')
   item = modulestore().get_item(usage_key)
   item.icon_override = 'video'
   modulestore().update_item(item, 1)
   ```
1. Go to [this page](http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/courseware/interactive_demonstrations/19a30717eff543078a5d94ae9d6c18a5/?child=first) and see that the first unit icon is now [fa-film](https://fontawesome.com/v4.7.0/icon/film).


**Reviewers**

**Settings**
```yaml
EDXAPP_EXTRA_REQUIREMENTS:
  - name: git+https://github.com/open-craft/custom-unit-icons.git@agrendalath/bb-933-address_override_changes#egg=custom-unit-icons
  - name: git+https://github.com/edx/edx-django-utils.git@master#egg=edx-django-utils

OVERRIDE_GET_UNIT_ICON: custom_unit_icons.icons.get_icon
XBLOCK_EXTRA_MIXINS:
    - custom_unit_icons.icons.IconOverrideMixin
```

**TODO**:
- [ ] Provision the sandbox.
- [x] Update the docs from #24057.
